File: src/main/java/de/lukaseichler/recurrentneuralnetwork/datatransformation/Scale.java
package de.lukaseichler.recurrentneuralnetwork.datatransformation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author leichler
 */
public class Scale {

    public static List<Double> scale(final List<Double> numbers) {
        final double max = Collections.max(numbers);
        final double min = Collections.min(numbers);
        final double range = max - min;
        List<Double> results = new ArrayList<>(numbers.size());

        results.addAll(numbers.stream().map(d -> (d - min) / range).collect(Collectors.toList()));
        Collections.sort(results);
        return results;
    }
}

File: src/main/scala/de/lukaseichler/recurrentneuralnetwork/datatransformation/Scale.scala
package de.lukaseichler.recurrentneuralnetwork.datatransformation

import scala.collection.mutable

/**
 * @author leichler
 */
object Scale {
  /**
   * Scales a list of double values onto a interval of [0,1]
   * @param numbers list of numbers to be scaled
   * @return scaled values
   */
  def scale(numbers: List[Double]): mutable.Seq[Double] = {
    val max = numbers.max
    val min = numbers.min
    val range = max - min
    val results = new mutable.MutableList[Double]

    results ++= numbers.map(x => (x - min) / range)
    results.sorted
  }
}


